EngineLayer/Proteomics/Protease.cs

```diff
@@ -87,9 +93,337 @@ internal List<int> GetDigestionSiteIndices(string sequence)
+                {
+                    intervals.Add(new ProteinInterval(1, protein.Length, protein, 0, "full"));
+                }
+
+                // cleave methionine
+                if ((initiatorMethionineBehavior != InitiatorMethionineBehavior.Retain && protein[0] == 'M') && OkayLength(protein.Length - 1, minPeptidesLength, maxPeptidesLength))
+                {
+                    intervals.Add(new ProteinInterval(2, protein.Length, protein, 0, "full:M cleaved"));
+                }
+
+                // Also digest using the proteolysis product start/end indices
+                foreach (var proteolysisProduct in protein.ProteolysisProducts)
+                {
+                    if (proteolysisProduct.OneBasedEndPosition.HasValue && proteolysisProduct.OneBasedBeginPosition.HasValue 
+                        && OkayLength(proteolysisProduct.OneBasedEndPosition.Value - proteolysisProduct.OneBasedBeginPosition.Value + 1, minPeptidesLength, maxPeptidesLength))
+                    {
+                        intervals.Add(new ProteinInterval(proteolysisProduct.OneBasedBeginPosition.Value, proteolysisProduct.OneBasedEndPosition.Value, protein, 0, proteolysisProduct.Type));
+                    }
+                }
+            }
+
+            // Cleavage rules for nonspecific search
+            else if (CleavageSpecificity == CleavageSpecificity.Semi)
+            {
+                intervals.AddRange(SemiProteolyticDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            // Full proteolytic cleavage
+            else if (CleavageSpecificity == CleavageSpecificity.Full)
+            {
+                intervals.AddRange(FullDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            else
+            {
+                throw new NotImplementedException();
+            }
+
+            return intervals;
+        }
+
+        /// <summary>
+        /// Gets the protein intervals based on nonspecific digestion rules
+        /// </summary>
+        /// <param name="protein"></param>
+        /// <param name="initiatorMethionineBehavior"></param>
+        /// <param name="maximumMissedCleavages"></param>
+        /// <param name="minPeptidesLength"></param>
+        /// <param name="maxPeptidesLength"></param>
+        /// <returns></returns>
+        private IEnumerable<ProteinInterval> SemiProteolyticDigestion(Protein protein, InitiatorMethionineBehavior initiatorMethionineBehavior, int maximumMissedCleavages, int? minPeptidesLength, int? maxPeptidesLength)
+        {
+            List<ProteinInterval> intervals = new List<ProteinInterval>();
+            List<int> oneBasedIndicesToCleaveAfter = GetDigestionSiteIndices(protein.BaseSequence);
+            for (int i = 0; i < oneBasedIndicesToCleaveAfter.Count - maximumMissedCleavages - 1; i++)
```

Comment:
@zrolfs Why didn't you loop through the missed cleavages like in the other methods?